### DISCUSSION TITLE
Optimal blocking rule for longer text fields such as full address or product description fields.

### DISCUSSION THREAD

**Discussion by seperman** [moj-analytical-services/splink#2004] [Q&A]
Hello,
Thanks for open-sourcing Splink.
What is the optimal blocking rule for longer text fields such as full address or product description fields? These fields may contain a few words or many words. So, a blocking rule simply based on exact matches or Levenshtein distance does not work.

For example, the dedupe library has the following predicates for short strings:

```
        predicates.commonFourGram,
        predicates.commonSixGram,
        predicates.tokenFieldPredicate,
        predicates.suffixArray,
        predicates.doubleMetaphone,
        predicates.metaphoneToken,
```

vs. long text:

```
        predicates.TfidfNGramCanopyPredicate,
        predicates.TfidfNGramSearchPredicate,
        predicates.TfidfTextCanopyPredicate,
        predicates.TfidfTextSearchPredicate,
```

I don't see, for example, how to use n-grams in Splink. The examples in the docs are all based on short strings. What do you recommend for long strings?
I am using the duckdb backend.
Thanks.

**Comment by RobinL** [comment]
It's a good question.  We don't have many good solutions for this at the moment, but here are some pointers.  You'd create the arrays in preprocessing (i.e. pre-importing to splink) using whatever algo you wanted (ngram or whatever).

- It is possible to [block on lists](https://github.com/moj-analytical-services/splink/pull/1692), but it takes quite a significant performance hit.  Unfortunately this is largely undocumented at the moment.  See [here](https://github.com/moj-analytical-services/splink/pull/1664#issuecomment-1777195458) for test script 
- Another hack is to create your array  and then have several blocking rules like `["l.my_arr[1] = r.my_arr[1]", "l.my_arr[1] = r.my_arr[2]", "l.my_arr[1] = r.my_arr[3]"` etc.  Obvs not ideal
- Another thing that occurred to me that would work quite nicely is using embeddings where you can [control the dimensionality](https://github.com/RobinL/cora_splink_embeddings/blob/76ec0ddb029e81de07160a52bb763f8f6f27a1c4/data_prep/01_process_raw_data.py#L27) (e.g. [text-embedding-3-small](https://openai.com/blog/new-embedding-models-and-api-updates).  These have the nice feature that the position matters, so unlike n-grams, you know the item at index 1 has the same 'meaning' on both sides.  You could reduce the dimensionality to some smallish number, then quantise the embeddings and block on them (e.g. `"l.my_embed[1] = r.my_embed[1] and l.my_embed[2] = r.my_embed[2]` etc.)

**Comment by seperman** [comment]
Hi @RobinL 
Thanks for getting back to me.

1. Is it possible to pass the thresholds for `levenshtein_at_thresholds` dynamically based on the length of 2 strings that are being compared? That way, we can adjust the threshold accordingly when blocking on a field with significant length differences between different rows.
2. For embeddings, can we use the cosine similarity function in Duckdb and set a threshold for blocking instead of comparing for exact matches in embedding?

**Reply by RobinL** [reply]
On (1), I'm not sure.  On (2), yes you could.   But unless your data is small it's probably not a good idea, because they're not equi-join conditions.  See here for an explainer
https://moj-analytical-services.github.io/splink/topic_guides/blocking/performance.html#efficient-blocking-rules

**Comment by lamaeldo** [comment]
This is quite late to the party but you may be interested in [RapidFuzz's Ratio and WRatio](https://github.com/rapidfuzz/RapidFuzz), they're quite clever ways to compare longer strings based on tokens (WRatio specifically if you are processing strings of varying length)

**Comment by seperman** [comment]
@lamaeldo That seems like a very expensive way to do blocking, no? It is O(n^2) since you have to compare all the pairs with each other.

**Comment by lamaeldo** [comment]
Oh absolutely, it would be terribly inneficient, but it still seems fast enough to work on smaller datasets. But I think it could be of most use in comparison levels
